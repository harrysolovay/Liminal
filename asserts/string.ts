import { assert } from "../util/mod.ts"

export function minLength(value: string, minLength: number) {
  assert(value.length >= minLength, `Length must be gte ${minLength} chars.`)
}

export function maxLength(value: string, maxLength: number) {
  assert(value.length <= maxLength, `Length must be lte ${maxLength} chars.`)
}
