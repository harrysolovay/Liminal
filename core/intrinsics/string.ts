import type { Rune } from "../Rune.ts"
import { Type } from "../Type.ts"

export const string: Rune<string, never> = Type("string", () => string)
