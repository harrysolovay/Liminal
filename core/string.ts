import type { Type } from "../Type.ts"
import { declare } from "../TypeDeclaration.ts"
import { assert } from "../util/assert.ts"

export const string: Type.Initial<string, {
  minLength: number
  maxLength: number
}> = declare<string>()({
  name: "string",
  source: {
    getType: () => string,
  },
  subschema: () => ({ type: "string" }),
  visitor: (value) => value,
  assertRefinementsValid: ({ minLength, maxLength }) => {
    assert(
      !(typeof minLength === "number" && typeof maxLength === "number") || minLength <= maxLength,
    )
  },
  assertRefinements: {
    minLength: (value, minLength) => assert(value.length >= minLength),
    maxLength: (value, maxLength) => assert(value.length <= maxLength),
  },
})
