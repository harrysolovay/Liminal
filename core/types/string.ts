import { Type } from "../Type.ts"

export const string: Type<string> = Type<string>({
  name: "string",
  source: {
    getType: () => string,
  },
})
