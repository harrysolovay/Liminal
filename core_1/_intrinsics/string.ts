import { Rune } from "../Rune.ts"
import type { Type } from "../Type/Type.ts"

export const string: Type<string> = Rune({
  kind: "Type",
  self: () => string,
})
