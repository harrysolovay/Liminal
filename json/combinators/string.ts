import { Type } from "../../core/mod.ts"

export const string: Type<string> = Type<string>({
  getType: () => string,
})
