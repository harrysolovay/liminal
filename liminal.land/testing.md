# Liminal Testing <Badge type="warning" text="beta" />
