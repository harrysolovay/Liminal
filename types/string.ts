import { declare, type Type } from "../core/mod.ts"
import { assert } from "../util/assert.ts"

export const string: Type.Initial<string, {
  minLength: number
  maxLength: number
}> = declare({
  name: "string",
  source: {
    getType: () => string,
  },
  refinements: ({ minLength, maxLength }) => {
    assert(
      !(typeof minLength === "number" && typeof maxLength === "number") || minLength <= maxLength,
      "`minLength` cannot be greater than `maxLength`.",
    )
    return {
      minLength: `Length must be gte ${minLength} chars.`,
      maxLength: `Length must be lte ${maxLength} chars.`,
    }
  },
  subschema: () => ({ type: "string" }),
  output: (f) =>
    f({
      refinementPredicates: {
        minLength: (value, minLength) => value.length >= minLength,
        maxLength: (value, maxLength) => value.length <= maxLength,
      },
    }),
})
