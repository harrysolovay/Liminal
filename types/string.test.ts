import { assertTypeSnapshot } from "testing"
import { string } from "./string.ts"

Deno.test("string", async (t) => {
  await assertTypeSnapshot(t, string)
})
