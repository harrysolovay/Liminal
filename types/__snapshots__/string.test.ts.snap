export const snapshot = {};

snapshot[`string 1`] = `
{
  type: "string",
}
`;

snapshot[`string 2`] = `
{
  description: "One.",
  type: "string",
}
`;

snapshot[`string 3`] = `
{
  description: "One. Two.",
  type: "string",
}
`;
